### ISSUE TITLE
Implement edit priority

### ISSUE THREAD

**Issue by cmyr** [xi-editor/xi-editor#933]
In certain instances, when applying a delta at a cursor location, the updated cursor location should _precede_ the inserted text. For instance, if the cursor is at EOL and the user types an open paren, a plugin should be able to insert a closing paren at the cursor location without it appearing _before_ the cursor.

I'm not sure what the actual implementation details of this are; @raphlinus will probably have to fill that in a little bit.

**Comment by nangtrongvuon** [comment]
I'm interested in working on this, since it'll solve #849 nicely. I'll appreciate all the help I can get from @raphlinus!

**Comment by raphlinus** [comment]
Here are my thoughts. It turned out to be quite a document :)

## Overview

The "priority" mechanism is required for correctness when using a CRDT-style mechanism for edits. Essentially, when there are two concurrent insertions at the same location, priority breaks the tie. This idea of priority for tiebreaking dates back to early Operational Transformation Literature (see section "an unsolved dOPT problem" in this [Operational Transformation] review paper).

A motivation for this issue is that the underlying priority mechanism can also be used for effects involving the interaction of insertion and the cursor. There's an example described in [Rope Science 10] which I'll recap here.

Let's say the buffer is in this state (we'll just use `|` to reprsent the caret position):

```
{|}
```

The user presses Enter, putting the buffer in this state:

```
{
|}
```

At this point, language plugin inserts whitespace, four spaces before the caret and a newline after:

```
{
    |
}
```

Concurrently, the user types "println". Let's say, though, that the plugin is slow, and the insertions are delayed. So, for a moment, the buffer is in this state:

```
{println|}
```

Finally, the plugin edits catch up. There are effectively three inserts at the caret: `"    "` at the lowest priority, `"println"` at the next priority, (implicitly the caret), and a newline at the last priority. So the final buffer is, as desired:

```
{
    println|
}
```

The goal is to allow the language plugin to be slow without causing text insertion to block on the calculation of automatically inserted whitespace. Of course, for this example it's not challenging to compute the desired whitespace quickly, but a major part of the motivation is to allow deeper analysis, for example calling out to an external tool such as gofmt or rustfmt.

It's likely that other interactions between insertion and selections can happen under the banner of priorities, but I haven't analyzed them carefully. I'm open to discussing them further in this issue. Potentially transpose could be redone so that transpose of "a|b" could be implemented as a deletion of "ab" and an insert of "a" and "b" with "b" before "a" in priority, and also with caret priority between them; there's a lot of discussion at #457 but I haven't had a chance to carefully review yet (though I do see that priorities are mentioned).

In order to make this pattern work, a bunch of things are needed, some of which already exist, some of which are still needed:

   [x] Annotate every insertion with a priority
   [x] Use the priority to order multiple concurrent insertions at the same point
   [x] Allow plugins to explicitly indicate a priority
   [ ] Have result of `commit_delta` track priority
   [ ] Use priority in selection transformation

## Priority is a pair: intent and tiebreaking

In order to encompass both editing and collaboration use cases, the priority must include both the intent (three different values in the above example, with a breakpoint so each possible value is ordered either before or after the caret) and the peer identifier. Thus, it's always possible to unambiguously order any two priorities, using a lexicographical order: if the intents are different, they decide, but in the case of two insertions with the same intent but from different peers, the peer id decides.

The `FullPriority` struct in `rope::Engine` already encodes this, and the idea of lexicographical ordering is very concisely encoded by the `#[derive(PartialOrd)]`.

A minor concern is that all edits on the same device currently get the same session id, even if they come from different plugins. I don't think this will affect correctness, because edits on the same device are sequentially ordered. I think we can revisit this question later.

Conflict resolution between two insertions should already be covered; this logic is in `Engine::mk_new_rev` and `Engine::rebase`. I'm pointing this out not because it's work to be done, but to reference.

## Interaction with selections

The core of the new work is to make the update of selections in response to an edit also sensitive to priority. Right now (I believe), the caret is always updated to be after a new insertion. (Something to investigate: what `keep_selections` does, which I believe is mostly set for transpose). This logic is in `View::after_edit`. The `Selection::apply_delta` function already takes an `after` parameter, currently hardcoded to `true`. So the main thing to be done is to plumb the priority from the original edit through to that.

One complication is that the code is currently architected so that there can be multiple `add_delta` operations, followed by a single `commit_delta`, then the delta is extracted from that. I see two ways to address this. The simpler but potentially less flexible is to track the priority separately in the editor state, and ensure that only inserts of the same priority are added in a single commit. You can see the problem of this for the purpose of transpose, in order to enforce this we'd need to do two commits. However, for plugin edits it shouldn't be a problem to generate a commit for each edit (but see below regarding how to handle multiple edits as in the whitespace insertion example above).

The more systematic approach is to track priority for each insertion in the engine, so that the delta resulting from `commit_delta` annotates each insertion in the resulting delta with a priority. That requires

Discussion question: why even separate add_delta from commit? What would go wrong if we did a commit after every delta? Originally a commit was tied to sending an update to the front-end, but maybe we can do multiple commits followed by a single update (i.e. there's an additional layer of defer in the minimal update logic, maybe we can exploit that).

## Handling multiple insert priorities from a plugin edit

The whitespace example above requires that the edit from the plugin include two different insertions with two different priorities. (The transpose example is similar, but is computed in core). There are a number of different ways to represent such an edit; here I'll discuss some pros and cons.

### One RPC, delta contains multiple insertions, priority per insertion

In the current code, a `PluginEdit` contains a delta and a priority, which is applied to all insertions in the delta. In this alternative, the signature of `RopeDelta` is changed so that each `DeltaElement::Insert` is also annotated with a priority.

One consequence of this is that an edit can be applied atomically; there are no intermediate states where some insertions have been applied but not others.

The major downside is the (nontrivial) change to the plugin interface; it requires explicit priorities when building the delta. It also absolutely requires that `commit_delta` be changed to track and report per-insertion priorities.

### One RPC, multiple (RopeDelta, priority) tuples

In the current code, a `PluginEdit` contains a single `RopeDelta` and a single priority. In this approach, the signature of `RopeDelta` doesn't change, but the `PluginEdit` is changed to accept a `Vec` of (`RopeDelta`, `priority`) pairs.

Similarly to above, multi-priority insertions are still atomic and there's no flicker to an intermediate state. Either `commit_delta` could track per-insertion priority, or a separate commit could be generated for each `RopeDelta` in the request.

There's another fundamental decision here, about see which below. Are the multiple edits based on the same base rev, or sequential? In the example above, the caret position after the user's newline is 2 (counting `"{\n"`). For both insertions to be based on the base rev, both the preceding whitespace and the following newline would be inserted at 2. For chaining, the following newline would be inserted at 6 (after the spaces), based on a rev that is the base rev plus the previous delta. The OT algorithms are a lot more complicated for the latter; if we want to keep the existing logic, it should be the same base rev.

Note also that chained deltas have more expressive power for what the delta can do. With chained deltas, let's say that starting from an empty buffer the first delta inserts "ab". Then the second delta can insert "x" at position 1, resulting in "axb". Requiring all insertions to be based on the same base rev means this can't be expressed; the insertion can be either before or after a previous insertion in the chain, but nothing else.

### Multiple RPCs, same base rev

This is more or less the status quo in the code. Each `PluginEdit` carries a delta with a single priority. For the whitespace example, the plugin sends two separate RPC's, both with the same base rev.

The display might flicker if it's updated between the two RPC's.

### Multiple RPCs, chained

This option is the combination of the last two. The potential advantage is that plugins have a reasonably simple conceptual model; they're basically editing their own local snapshot of the document, and can do basically unrestricted edits over an arbitrarily long time.

The downside is that the OT logic becomes a _lot_ more complicated.

### Multiple RPCs, each waiting on ack of the previous one

This alternative is only for illustration; it places a ridiculously high burden on plugin authors, and the flickering problem is likely to be much worse. In this alternative, there are two RPC's sent, and the second is based on the rev sent by the core as an ack from the first.

Of course, if there are other edits merged concurrently, then this means that the plugin has to track those edits, fix up the delta based on OT transformation, and send that. This is extremely complicated. The only real advantage is that this can be implemented without any changes to core.

### Discussion

I personally lean toward the 3rd choice here, because it's the simplest to implement. If the flickering is a problem, we can migrate towards choice 2, also without much disruption.

However, it's worth considering. How easy is it to generate the required deltas? In the whitespace case? In the transpose case? Is the API comprehensible to plugin authors?

Note that for syntax highlighting we definitely want to stream edits based on the same base rev. The idea of the plugin API is that it supports both use cases using the same general mechanism.

Having written this, it's clear we should look at the question of "commit." The main function of commit is to update the style spans and the selection in response to a delta. Right now, there's a decoupling; we can add multiple deltas, have the CRDT engine synthesize a delta that aggregates them, and then the commit is based on that. Maybe we should couple more tightly, so that every edit is immediately commited; but multiple commits can be aggregated into one update.

Some edits effectively come from the CRDT (primarily undo/redo), rather than from direct edits from either front-end or plugins. When we get a delta this way, we currently lose some information to allow reconstruction of the selection state (see #765). I think that's a separate issue, but overlaps with the concerns above.

These feel like deep issues. I'm happy to discuss and answer questions that arise.

[Rope Science 10]: http://xi-editor.github.io/xi-editor/docs/rope_science_10.html
[Operational Transformation]: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.53.933&rep=rep1&type=pdf

**Comment by nangtrongvuon** [comment]
@raphlinus That was a very informative (and quite easy, for me at least, to follow) post! 

I'm going to base what I'm about to write on my little experience trying to implement a line move while working with the existing implementation of priorities. I think I agree with you that sending multiple edit RPCs with the same base rev would make the most sense, since it'd tie nicely to an undo for that operation as well.

I ran into this problem early on while trying to implement a line move:

The file could start looking something like this:
`
{
let line1 = "line1";
let line2 = "line2";
}|
`

At that current caret position, the user moves that line up. This should then look like:
`
{
let line1 = "line1";
}|
let line2 = "line2";
`

This is represented by moving the line *above* down. The delta looks like "cut the line above, then paste it after the line that we wanted to move", which would mirror an actual "line move" movement where a user would cut a line then paste it where they want it to be.

However, current Xi would resolve this (since caret is always updated to be after an insertion) as:
`
{
let line1 = "line1";
}
let line2 = "line2";|
`
In this case, what I imagine to be the simplest solution is to have the deltas specify their desired caret position with a priority, and aggregate them into a complete delta that we then apply. This plays well with edits that require deltas which have differing priorities (such as the move line case above, or transposes, or appending texts), and also lets the API stay relatively the same. Also, not every edit has to be complicated enough to want to specify a priority, so I think options 1 (and perhaps 2 in a way) are a bit overkill.

I don't think generating deltas would be too hard in this case, or at least for me - even if move lines was to be move x lines up instead of just 1 line, it wouldn't be too hard to create the deltas for the desired edit. 

What I'm not sure here is how priorities will play with Undo and Redo. They have to know of the deltas' priorities too, right?

> Originally a commit was tied to sending an update to the front-end, but maybe we can do multiple commits followed by a single update (i.e. there's an additional layer of defer in the minimal update logic, maybe we can exploit that).

I think this could be done and it makes a lot of sense too, from my thoughts above. I'm all for implementing this, if we go with that multiple commit -> single update route.

> The more systematic approach is to track priority for each insertion in the engine, so that the delta resulting from commit_delta annotates each insertion in the resulting delta with a priority. That requires

Don't leave us hanging! :p

**Comment by raphlinus** [comment]
Just a few quick thoughts while this is fresh in my mind.

Moving lines up and down is _very_ much like transpose. I think in this scheme, "move line up" is best implemented as deleting the line above, and inserting it at the caret with priority "after caret." I believe that would also work in the case where the selection is a block of lines - it would leave the selection untouched. So it sounds like this is a good use case for priorities.

I want to clarify the interaction with undo/redo a bit. From the point of view of the CRDT engine, priorities are used to resolve the merge of multiple concurrent edits. Once the merge is resolved, the edit is just a straightforward edit on the string. So undo just reverses that delta, and it doesn't really matter what priorities created that edit.

The trickier aspect is how these deltas derived from undo/redo interact with the selection. Right now, if the delta inserts at the caret, it just moves the caret after the insertion. This is not ideal, it would be nice to restore the selection state as well. I'm inclined to think that the most systematic solution is to track the selection state in the CRDT as well. This also touches the ongoing discussion in #913. But I think we can largely put these questions aside as being orthogonal to priorities.

> Don't leave us hanging! :p

Heh, edited in place.

**Comment by brainkim** [comment]
> A minor concern is that all edits on the same device currently get the same session id, even if they come from different plugins. I don't think this will affect correctness, because edits on the same device are sequentially ordered. I think we can revisit this question later.

I'm working on an independent implementation of the xi editor crdt and I'm wondering if it's better to panic if the engine encounters an equal priority in `Engine::edit_rev`? Currently there's just a comment saying this should never be the case. https://github.com/xi-editor/xi-editor/blob/master/rust/rope/src/engine.rs#L352

**Comment by cmyr** [comment]
A panic would be completely reasonable, if two revs are equal we are in trouble. If you look at https://github.com/xi-editor/xi-editor/blob/master/rust/rope/src/engine.rs#L79 you can see some napkin math on the probabilities (tl dr: a collision is unlikely.)

**Comment by brainkim** [comment]
@cmyr In my implementation I'm using uuids, so I'm not so much worried about the client id collisions as I am worried about plugins writing concurrent edits with the same priority on the same client (possibly due to an oversight by the plugin author). I've only really looked at the rope/crdt in this project so maybe there's some logic in the plugin code which prevents this?

**Comment by cmyr** [comment]
mm, interesting, missed that. I'm not sure how much of a problem this is, but one option would be to use plugin id as a tiebreaker.

**Comment by lord** [comment]
@cmyr 

So I spent some time thinking about how adjacent selections could possibly integrate with edit priority. The problem (from #647) looks like this, with two adjacent selections.

    |text|here|
    ^    ^       selection A
         ^    ^  selection B

Let's say two plugins both do a surround edit, one plugin wrapping `[]` with high priority, and the other wrapping `{}` with a low priority. We'd expect, no matter which plugin gets to the insert first, for it to look like:

    {[|text|]}{[|here|]}
      ^    ^           selection A
                ^    ^ selection B

The problem I'm struggling with is how does the engine order these inserts and cursors, which all happen at exactly the same point:

- selection A end
- selection B start
- insert `}` and then `{` at low priority
- insert `]` and then `[` at high priority

In particular, knowing to put `}{` inside of `][` I think is tricky? I considered a couple ideas, including using 'before' and 'after' cursor information to figure out which selection an insert was part of, but I think the solution I came up with didn't play well with the case of adjacent selections being transposed.

Perhaps a better solution could be associating which cursor is responsible for which insert? That could simplify the case of adjacent cursors for a variety of edits, and then the cursor priority could be simpler, something more like before-inside-after instead of a four segment option?

As a disclaimer, still don't fully understand priority, but I also wonder if the priority information (or at least its InsertDrift aspect) should be on a per-insert basis, instead of a per-delta basis like InsertDrift currently is? Feels a little weird to have to do two separate Deltas for auto closing pairs (typing `(` to get `(|)` with a cursor in the middle),
